opentelemetry-api/src/opentelemetry/correlationcontext/__init__.py

```diff
@@ -0,0 +1,128 @@
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import copy
+import itertools
+import string
+import typing
+from contextlib import contextmanager
+
+from opentelemetry.context import ContextAPI
+
+CORRELATION_CONTEXT_KEY = "correlation-context-key"
+
+
+class CorrelationContextManager:
+    """
+    Manages access to CorrelationContext entries
+    """
+
+    def __init__(
+        self, entries: typing.Optional[typing.Dict[str, typing.Any]] = None
+    ) -> None:
+        if entries:
+            self._correlation_context = copy.deepcopy(entries)
+        else:
+            self._correlation_context = {}
+
+    def get_correlations(self) -> typing.Dict:
+        """
+        Returns the entries in this CorrelationContext. The order of entries is not significant. Based
+        on the languagespecification, the returned value can be either an immutable collection or an
+        immutable iterator to the collection of entries in this CorrelationContext.
+
+        Returns:
+            An immutable iterator to the collection of entries in this CorrelationContext
+        """
+        return self._correlation_context.items()
+
+    def get_correlation(
+        self, context: ContextAPI, name: str
+    ) -> typing.Optional[typing.Any]:
+        """
+        To access the value for an entry by a prior event, the Correlations API provides a
+        function which takes a context and a name as input, and returns a value. Returns
+        the Value associated with the given Name, or null if the given Name is not present.
+
+        Args:
+            context: The context in which to find the CorrelationContext
+            name: The name of the entry to retrieve
+        Returns:
+            The value of the entry matching the name
+        """
+        # pylint: disable=no-self-use
+        correlation_context = context.get_value(CORRELATION_CONTEXT_KEY)
+        if correlation_context and name in correlation_context:
+            return correlation_context[name]
+        return None
+
+    def set_correlation(
+        self, context: ContextAPI, name: str, value: typing.Any
+    ) -> ContextAPI:
+        """
+        To record the value for an entry, the Correlations API provides a function which takes a
+        context, a name, and a value as input. Returns an updated Context which contains the new value.
+
+        Args:
+            context: The context in which to find the CorrelationContext
+            name: The name of the entry to set
+            value: The value of the entry
+        Returns:
+            A new Context object with updated correlations
+        """
+        new_correlation_context = self._copy()
+        new_correlation_context[name] = value
+        return context.set_value(
+            CORRELATION_CONTEXT_KEY, new_correlation_context
+        )
```

Comment:
If I understand this, a `CorrelationContextManager` is a snapshot of the correlation context data of a `ContextAPI`, with manager-style helper methods for setting that data.

But since this isn't associated with a context instance, this makes for a pretty weird API! It's surprising that `set_correlation` and `remove_correlation` don't just set/remove values from the context or this snapshot, they replace the _entire_ context with values from the time the `CorrelationContextManager` was created.